pgjdbc/src/main/java/org/postgresql/largeobject/BlobInputStream.java

```diff
@@ -86,33 +86,74 @@ public BlobInputStream(LargeObject lo, int bsize, long limit) {
   public int read() throws java.io.IOException {
     LargeObject lo = getLo();
     try {
-      if (limit > 0 && apos >= limit) {
+      if (limit > 0 && absolutePosition >= limit) {
         return -1;
       }
-      if (buffer == null || bpos >= buffer.length) {
-        buffer = lo.read(bsize);
-        bpos = 0;
+      // read more in if necessary
+      if (buffer == null || bufferPosition >= buffer.length) {
+        buffer = lo.read(bufferSize);
+        bufferPosition = 0;
       }
 
       // Handle EOF
-      if (buffer == null || bpos >= buffer.length) {
+      if ( buffer == null || bufferPosition >= buffer.length ) {
         return -1;
       }
 
-      int ret = (buffer[bpos] & 0x7F);
-      if ((buffer[bpos] & 0x80) == 0x80) {
-        ret |= 0x80;
-      }
+      int ret = (buffer[bufferPosition] & 0xFF);
 
-      bpos++;
-      apos++;
+      bufferPosition++;
+      absolutePosition++;
 
       return ret;
     } catch (SQLException se) {
       throw new IOException(se.toString());
     }
   }
 
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    int bytesCopied = 0;
+    LargeObject lo = getLo();
+
+    /* check to make sure we aren't at the limit
+    *  funny to test for 0, but I guess someone could create a blob
+    * with a limit of zero
+    */
+    if ( limit >= 0 && absolutePosition >= limit ) {
+      return -1;
+    }
+
+    try {
+      // have we read anything into the buffer
+      if ( buffer != null ) {
+        // now figure out how much data is in the buffer
+        int bytesInBuffer = buffer.length - bufferPosition;
+        // figure out how many bytes the user wants
+        int bytesToCopy = len > bytesInBuffer ? bytesInBuffer : len;
+        // copy them in
+        System.arraycopy(buffer, bufferPosition, b, off, bytesToCopy);
+        // move the buffer position
+        bufferPosition += bytesToCopy;
+        // increment offset
+        off += bytesToCopy;
+        // decrement the length
+        len -= bytesToCopy;
+        bytesCopied = bytesToCopy;
+      }
+      if (len > 0 ) {
+        bytesCopied += lo.read(b, off, len);
+        absolutePosition += bytesCopied;
+        if ( bytesCopied == 0 && (buffer == null || bufferPosition >= buffer.length) ) {
```

Comment:
At this point, how could the second half of this condition ever be false?

Reply:
ah, pretty sure that should have been buffer != null thx, for the second set of eyes

Reply:
I think bytesCopied == 0 is the only thing to check.

Reply:
Pretty sure one of the tests failed, let me check again

Reply:
testGetBinaryStreamWithBoundaries fails without the second test

Reply:
Is the test actually right?
 I am not seeing how to get here with data remaining in the buffer.

Reply:
I more or less copied it from the read() function. You may be right that it will never be exercised.